Digital payment 2

anjana-bl opened the pull request.

Vishali24 committed:
Code changes to enable digital payment integration

Vishali2499 committed:
updated version in the file

Vishali2499 committed:
Done file formatting

Vishali2499 committed:
Added a new line at eof

Vishali2499 committed:
build and formatting files

anjana-bl committed:
fixing error TS2345

fixing error TS2345 in reducer function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SAP/spartacus?pullRequest=12749) <br/>All committers have signed the CLA.

anjana-bl committed:
resolving conflicts in package.json

resolving conflicts in package.json

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Delete tslint.json

Delete tslint.json

anjana-bl committed:
changing from tslint->eslint for digital-payments

changing from tslint->eslint for digital-payments

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Update occ-digital-payments.adapter.ts

- Using getBaseUrl() instead of getBaseEndpoint()

anjana-bl committed:
Update dependencies.json

update dependencies.json for digital-payments

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Update dependencies.json

anjana-bl committed:
Update dp-checkout-payment-request.reducer.ts

fixing errors during linting "digital-payments"

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Update dp-local-storage.service.ts

fixing ts2322

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
correcting errors during validating code formatting (using prettier)

correcting errors during validating code formatting (using prettier)

pratikkumar90 committed:
Merge branch 'develop' into digital-payment-2

marlass committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Update package.json

Update package.json

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

AshreeJay committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
fixing functionality errors

fixing functionality errors

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
changes related to checkout library

changes related to checkout library

anjana-bl committed:
changes for missing peerDependencies

contains  changes for Missing `@spartacus/checkout` dependency that is directly referenced in library.

anjana-bl committed:
fixing lint errors

fixing lint errors

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
Update dp-payment-method.component.ts

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
fixing ts2345

fixing ts2345

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
setting digital-payments to false by default

setting digital-payments to false by default

anjana-bl committed:
digital-payment e2e test script

digital-payment e2e test script

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Update cypress.ci.digital-payments.json

anjana-bl committed:
Update cypress.ci.digital-payments.json

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
updating api_url for digital-payment

updating api_url for digital-payment

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
changing user.ts

changing user.ts

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
for e2e digital-payment test script

for e2e digital-payment test script

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
fixing error TS2559

 fixing error TS2559

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
config update

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

marlass commented on .travis.yml:
```diff
@@ -37,6 +37,8 @@ jobs:
   - script: "export CYPRESS_CONFIG=cypress.ci.cdc.json && export SPA_ENV='cdc' && ./ci-scripts/e2e-cypress.sh -s cdc || exit 0"
     name: "E2E tests for CDC"
+  - script: "export CYPRESS_CONFIG=cypress.ci.digital-payments.json && export SPA_ENV='digital-payments' && ./ci-scripts/e2e-cypress.sh -s digital-payments || exit 0"
+    name: "E2E tests for CDC"   
```
```suggestion
    name: "E2E tests for Digital Payments"   
```

marlass commented on integration-libs/digital-payments/.release-it.json:
```diff
@@ -0,0 +1,34 @@
+    "publishPath": "./../../dist/digital-payments"
+  },
+  "hooks": {
+    "after:version:bump": "cd ../.. && ng build digital-payments --prod"
```
```suggestion
    "after:version:bump": "cd ../.. && ng build digital-payments --configuration production"
```

marlass commented on integration-libs/digital-payments/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@spartacus/digital-payments",
+  "version": "0.300.0",
+  "description": "Customer Data Cloud Integration library for Spartacus",
```
```suggestion
  "description": "Digital Payments Integration library for Spartacus",
```

marlass commented on integration-libs/digital-payments/src/checkout/adapters/occ-digital-payments.adapter.ts:
```diff
@@ -0,0 +1,47 @@
+  createPaymentRequest(
+    userId = CURRENT_USER,
+    cartId = CURRENT_CART
+  ): Observable<any> {
```
Can we use type `unknown` instead of `any`?

anjana-bl replied:
changing it to unknown was giving ts(2416), so i changed it to 'DpPaymentRequest'.

marlass commented on integration-libs/digital-payments/tsconfig.lib.prod.json:
```diff
@@ -0,0 +1,6 @@
+  "extends": "./tsconfig.lib.json",
+  "angularCompilerOptions": {
+    "enableIvy": false
+  }
```
```suggestion
  "compilerOptions": {
    "declarationMap": false
  },
  "angularCompilerOptions": {
    "compilationMode": "partial"
  }
```

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

anjana-bl committed:
error during build

fixing - Can not find module “@angular-devkit/build-angular”

anjana-bl committed:
Update angular.json

anjana-bl committed:
Update .travis.yml

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>

anjana-bl committed:
Update integration-libs/digital-payments/package.json

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>

anjana-bl committed:
Update integration-libs/digital-payments/.release-it.json

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>

anjana-bl committed:
Update integration-libs/digital-payments/tsconfig.lib.prod.json

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>

marlass commented on scripts/changelog.ts:
```diff
@@ -325,6 +325,10 @@ if (typeof config.to === 'undefined') {
     case '@spartacus/cdc':
       config.library = '@spartacus/cdc';
       break;
+    case 'digital-payments':
```
Could you also add digital payments to `changelog.ejs` file after @spartacus/cdc?

anjana-bl replied:
done

anjana-bl committed:
Update occ-digital-payments.adapter.ts

marlass committed:
chore: Add user lib paths to tsconfig

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

marlass commented on projects/storefrontapp-e2e-cypress/cypress/integration/vendor/digital-payments/digital-payments.e2e-spec.ts:
```diff
@@ -0,0 +1,30 @@
+  reviewAndPlaceOrder,
+} from '../../../helpers/vendor/digital-payments/user';
+
+describe('checkout using digital-payments', () => {
```
There should be another e2e test checking the process of adding new credit card.

anjana-bl replied:
The new credit card is filled in a 3rd party page, with a completely different domain. Cypress.io cannot navigate across different domains, so the test script is written this way.

anjana-bl replied:
Since cypress.io cannot navigate from one domain to another, the 3rd party page navigation is not possible. Hence the test script always uses a defaulted credit card.

marlass commented on integration-libs/digital-payments/src/config/digital-payments-config.ts:
```diff
@@ -0,0 +1,18 @@
+  providedIn: 'root',
+  useExisting: Config,
+})
+export abstract class DigitalPaymentsConfig {
```
Where this config is used?

anjana-bl replied:
its unused, so removed it.

marlass commented on integration-libs/digital-payments/src/checkout/store/actions/dp-checkout-payment-details.action.ts:
```diff
@@ -0,0 +1,50 @@
+import { DP_CHECKOUT_PAYMENT_DETAILS } from '../digital-payments-state';
```
For new features we stopped using ngrx as a way to cache responses and invoke API calls.

Instead queries and commands should be used. Example of feature leveraging that is new user library.

From what I've seen in this PR this would make the whole integration library much smaller, as almost half of it is just ngrx boilerplate.

marlass commented on integration-libs/digital-payments/src/checkout/cms-components/dp-payment-method/dp-payment-form/dp-payment-form.component.html:
```diff
@@ -0,0 +1,4 @@
+<div class="text-center">{{ 'dpPaymentForm.redirect' | cxTranslate }}</div>
+<div class="cx-spinner"><cx-spinner></cx-spinner></div>
+
+<ng-template #loading> </ng-template>
```
Is it used?

anjana-bl replied:
yes, this is used to show the message 'You will soon be redirected to third party page for card registration' before going to 3rd party page

marlass commented on integration-libs/digital-payments/src/checkout/cms-components/dp-payment-method/dp-payment-callback/dp-payment-callback.module.ts:
```diff
@@ -0,0 +1,15 @@
+@NgModule({
+  imports: [CommonModule, SpinnerModule, I18nModule],
+  declarations: [DpPaymentCallbackComponent],
+  entryComponents: [DpPaymentCallbackComponent],
```
Since ng12 update we no longer use `entryComponents` in modules.

anjana-bl replied:
done

marlass commented on integration-libs/digital-payments/src/checkout/cms-components/dp-payment-method/dp-payment-callback/dp-payment-callback.component.ts:
```diff
@@ -0,0 +1,72 @@
+  paymentDetailsAdded = new EventEmitter<any>();
+
+  constructor(
+    private dpPaymentService: DpCheckoutPaymentService,
+    private dpStorageService: DpLocalStorageService,
+    private globalMsgService: GlobalMessageService,
+    private route: ActivatedRoute
```
For extensibility we use `protected` for contructor parameters

anjana-bl replied:
done

marlass commented on integration-libs/digital-payments/src/checkout/adapters/occ-digital-payments.adapter.ts:
```diff
@@ -0,0 +1,48 @@
+    cartId = CURRENT_CART
+  ): Observable<DpPaymentRequest> {
+    return this.http.post<any>(
+      `${this.occEndpoints.getBaseUrl()}/users/${userId}/carts/${cartId}/payment/digitalPayments/request`,
```
Instead of hardcoding url here it should be defined in `OccEndpoints` and it should be fetched with `buildUrl` passing onlu the required params (userId, cartId).

Applies to second method as well.

anjana-bl replied:
changes are done.

marlass reviewed (changes requested):
Also I recommend adding schematics to this library, so it can be installed by customer fully automatically.

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
changes for review comments

anjana-bl committed:
Merge branch 'digital-payment-2' of https://github.com/SAP/spartacus into digital-payment-2

marlass committed:
chore: Update tsconfig tool to add nested spartacus dependencies

marlass committed:
chore: Update config

marlass committed:
Merge branch 'digital-payment-2' of github.com:SAP/spartacus into digital-payment-2

anjana-bl committed:
using buildUrl

anjana-bl committed:
Merge branch 'develop' into digital-payment-2

anjana-bl committed:
removing unused config

marlass commented:
Merged with other PR
